backtrace/test/crashpad_utils.rb

```diff
@@ -0,0 +1,143 @@
+        unless defined?(DEBUG)
+            webrick_params = webrick_params.merge({
+                Logger: WEBrick::Log.new(OS.windows? ? "NUL:" : "/dev/null"),
+                AccessLog: []
+            })
+        end
+        server = WEBrick::HTTPServer.new webrick_params
+        ret = nil
+
+        server.mount_proc '/alive' do |req, res|
+            res.body = ''
+            res.status = 200
+        end
+
+        server.mount_proc '/crashpad_upload' do |req, res|
+            req.body
+            res.body = 'Finished?'
+            res.status = 200
+            ret = [req, res]
+            server.shutdown
+        end
+
+        server.start
+
+        ret
+    end
+
+    def await_availability timeout = 1
+        client = HTTPClient.new
+        b = Time.now
+        while (Time.now - b < timeout) do
+            begin
+                r = client.get('http://localhost:6097/alive')
+                return true if r.code == 200
+            rescue Errno::ECONNREFUSED => e
+            end
+        end
+        false
+    end
+end
+
+class Upload
+    attr_reader :req, :res, :params
+
+    def initialize req, res
+        @req = req
+        @res = res
+        parse_form_data
+    end
+
+    def parse_form_data
+        body = case @req.header.fetch('content-encoding', nil)
+        when ['gzip']
+            Zlib::gunzip @req.body
+        else
+            @req.body
+        end
+        boundary = @req.content_type.scan(/boundary=(.*)/)[0][0]
+        @params = WEBrick::HTTPUtils::parse_form_data body, boundary
+    end
+end
+
+class Execution
+    # expect to be called from [git_root]/backtrace/test by default
+    def initialize build_dir: '../../cbuild/', url: DEFAULT_URL, executable:
+        @build_dir = build_dir
+        @executable = executable
+        @url = url
+    end
+
+    def execute tmp_dir: nil
+        Dir.chdir @build_dir do
+            execute = lambda do |tmp|
+                argv = [ @executable, @url, handler, tmp ]
+
+                out_r, out_w = IO.pipe
+                err_r, err_w = IO.pipe
+
+                pid = spawn *argv, out: out_w, err: err_w
+
+                out_w.close
+                err_w.close
+
+                Process.wait pid
+
+                {
+                    stdout: out_r.read,
+                    stderr: err_r.read
+                }
+            end
+
+            if tmp_dir
+                execute.call tmp_dir
+            else
+                Dir.mktmpdir { |dir| execute.call dir }
+            end
+        end
+    end
+
+    def handler
+        if OS.windows?
+            'handler/Debug/handler.exe'
+        else
+            'handler/handler'
+        end
+    end
+end
+
+def Crashpad.perform_test opts = {}
+    cl = Listener.new
+    ce = Execution.new **opts
+    execution_result = nil
+    thread = Thread.new do
+        return nil unless cl.await_availability
+        execution_result = ce.execute
+    end
+    upload = Upload.new *cl.listen_one
```

Comment:
I do not know ruby syntax, but I believe in line 131 you create a new thread to wait until rest API will be available, at the same time, you start the rest API later. 

1. listen_one doesn't look like async API, rather it looks like a synchronous API. wouldn't ie be better to start it first and then assume the rest API is available?
2. if we need to wait, wouldn't it be better to first start rest API and then create a thread?

Reply:
Precisely because it is synchronous, once I start listening, I will not get the control back until the listening is done (a crash is uploaded). That is why the thread waits for the availability of the server and then executes the Crashpad binary, which uploads the crash.